TodoList/sources/TodoList.Api/Controllers/NodesController.cs

```diff
@@ -5,40 +5,161 @@
 using TodoList.Contracts.Api;
 using TodoList.Contracts.Models;
 using TodoList.Contracts.Repository;
+using TodoList.Contracts.Services;
+using static System.String;
 
 namespace TodoList.Api.Controllers
 {
     [Route("api/v1/nodes/{id:guid?}", Name = "nodes")]
     public class NodesController : ApiController
     {
         private readonly INodesRepository _repository;
-        private readonly ILocationHelper _locationHelper;
+        private readonly ICreateNodeService _createNodeService;
+        private readonly IUpdateNodeService _updateNodeService;
+        private readonly ILocator _locator;
 
-        public NodesController(INodesRepository repository, ILocationHelper locationHelper)
+        public NodesController(
+            INodesRepository repository,
+            ICreateNodeService createNodeService,
+            IUpdateNodeService updateNodeService,
+            ILocator locator)
         {
             _repository = repository;
-            _locationHelper = locationHelper;
+            _createNodeService = createNodeService;
+            _updateNodeService = updateNodeService;
+            _locator = locator;
         }
 
         public async Task<IHttpActionResult> GetAsync()
             => Ok(await _repository.GetAllAsync());
 
         public async Task<IHttpActionResult> GetAsync(Guid id)
-            => Ok(await _repository.GetByIdAsync(id));
+        {
+            ValidateId(id);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var existingNode = await _repository.GetByIdAsync(id);
+
+            if (existingNode == null)
+            {
+                return NotFound();
+            }
+
+            return Ok(existingNode);
+        }
 
-        public async Task<IHttpActionResult> PostAsync(NodeModel model)
+        public async Task<IHttpActionResult> PostAsync([FromBody] NodeModel node)
         {
-            var returnedModel = await _repository.AddAsync(model);
-            return Created(_locationHelper.GetNodeLocation(returnedModel.Id), returnedModel);
+            ValidatePostNodeModel(node);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var newNode = await _createNodeService.CreateNodeAsync(node);
+
+            return Created(_locator.GetNodeLocation(newNode.Id), newNode);
         }
 
-        public async Task<IHttpActionResult> PutAsync(NodeModel model)
-            => Content(HttpStatusCode.Accepted, await _repository.UpdateAsync(model));
+        public async Task<IHttpActionResult> PutAsync([FromBody] NodeModel node)
```

Comment:
Sorry to intervene, but while it is completely correct suggestion to create a view model and it would actually be a correct solution, for the educational purses (~ to see how it sucks, both in API and connected front-end): validate in the controller that the `id` in model is either same as the one in URL or (~xor) require the `id` to be empty (as it has to be in `POST`)